docs/building.md

```diff
@@ -0,0 +1,176 @@
+# Cloning and building
+
+Fobnail project consists of 3 main repositories with code:
+
+- [Fobnail firmware](https://github.com/fobnail/fobnail) (Rust)
+- [Platform Owner](https://github.com/fobnail/fobnail-platform-owner) (C)
+- [Attester](https://github.com/fobnail/fobnail-attester) (C)
+
+There is also [Fobnail SDK](https://github.com/fobnail/fobnail-sdk) that
+simplifies building of firmware by using Docker container, but only
+[one script](https://github.com/fobnail/fobnail-sdk/blob/main/run-fobnail-sdk.sh)
+from this repository is actually needed.
+
+Following steps were successfully performed on Ubuntu 22.04.1 LTS.
+
+## Fobnail Token firmware
+
+Start by installing prerequisites and adding current user to `docker` group:
+
+```shell
+sudo apt -y install git docker.io
+sudo usermod -aG docker $USER
+```
+
+Log out of the desktop session then log in again - this is required after making
+the change to user groups. You may confirm it by running `groups`.
+
+Clone the repository along with its submodules:
+
+```shell
+git clone https://github.com/fobnail/fobnail --recurse-submodules
+```
+
+Build process for Token firmware is by far the most complicated one. For that
+reason Fobnail SDK is strongly suggested. To use it, simply download script
+`run-fobnail-sdk.sh` either by cloning whole repository or just this one script
+and save it to a directory in your `PATH` environment variable:
+
+```shell
+sudo wget -O /usr/local/bin/run-fobnail-sdk.sh \
+  https://raw.githubusercontent.com/fobnail/fobnail-sdk/main/run-fobnail-sdk.sh
+sudo chmod +x /usr/local/bin/run-fobnail-sdk.sh
+```
+
+With the script installed, one can prepare for the build process itself.
+Firmware can be build for a physical Fobnail Token or simulated on PC, with
+different preparatory steps required. In both cases, first execution of
+`build.sh` downloads Docker container image, which can be both time- and
+bandwidth-consuming.
+
+#### Environment variables common for both targets
+
+Fobnail Token firmware is configured with environment variables passed to
+`build.sh`.
+
+Note that due to the way Docker mounts directories all files and directories
+pointed to by following variables must be located somewhere in `fobnail`
+directory.
+
+- `FOBNAIL_PO_ROOT` - **required** option, must point to valid PEM or DER file
+  with Platform Owner's root certificate. See [this document](/keys_and_certificates/#platform-owner-certificate-chain)
+  for description of PO certificate chain and instructions for building such.
+
+- `FOBNAIL_EK_ROOT_DIR` - points to directory with TPM root certificates.
+  Fobnail repository includes [such directory](https://github.com/fobnail/fobnail/tree/main/tpm_ek_roots)
+  which normally should be used as `FOBNAIL_EK_ROOT_DIR`, but you may change it
+  if certificates supplied in repository became outdated. Either this or
+  `FOBNAIL_EXTRA_EK_ROOT` (or both) **must be specified**, but build process
+  won't fail when neither is passed.
```

Comment:
This is explained in very convoluted way, users need simple instruction what to do when. It is hard to make decision here what action I should take. IIUC:
* I have to identify what TPM I have
* I have to provide at least one valid TPM EK Root Cert for further verification of information provided by TPM
* `tpm_ek_roots` may contain TPM EK Root Cert for TPM I have, but checking validi of such certificate is left to user/developer because we cannot be responsible for tracking all vendors certs at this point of project

Please let me know if this valid reasoning.

What I don't understand:
* why we have to ways of providing EK Root Cert by file and by directory?
* why we not ask user/developer to improve content of `tpm_ek_roots` directory according to their needs, including emulator certs, for some use cases may just one vendor cert will be needed
* why we don't have TPM HCL based on that list? What would be use case of Fobnail for platform with not supported TPM?
* build process does not fail when no TPM EK Root Cert was provided? What would be use of such build? Maybe we should fail in such scenario?

Reply:
> We don't want to have such certificates in the tree. FOBNAIL_EXTRA_EK_ROOT was created to easily add one additional root certificate in this case, without messing with "official" root certificates.

So it means it is advanced option needed only in some cases. Knowledge of that option is not needed here, it can be described in place where we really use emulator. Correct?

> Not enough testing, for now HCL would consist of just one TPM.

Let's make it easy to extend HCL as Qubes and Dasharo try to do. I have various TPMs in various devices, what would be easiest way to provide feedback? We should let potential users know about that.

> If you're asking what Fobnail can be used for on platform that doesn't have its TPM's root certificate trusted, then it wouldn't even pass platform provisioning. Whether that's good or bad depends on threat model.

I think this meaning of second part of my question is closer to what I meant. I wonder why we pass compilation if no TPM EK Root Cert was provided. It sounds like Fobnail is useless for platforms for which we don't know TPM EK Root Cert. In such situation we should fail compilation without at least on TPM EK Root Cert provided.

Reply:
> So it means it is advanced option needed only in some cases. Knowledge of that option is not needed here, it can be described in place where we really use emulator. Correct?

Yes and no, I guess this connects with Diataxis, we can either have simple instructions or all options listed in one place. I'm not sure which option is more advanced use case though, there may be a lot of users that don't have TPM2.0 and would prefer to use emulated TPM for testing.

> Let's make it easy to extend HCL as Qubes and Dasharo try to do. I have various TPMs in various devices, what would be easiest way to provide feedback?

GH issues maybe? We would have to prepare a process and a template, probably with some specific logs and maybe certificate chains.

Reply:
https://github.com/fobnail/fobnail/pull/61 should make it fail with no EK root is passed. Documentation was updated accordingly. Now CI fails as expected :slightly_smiling_face:

Reply:
> Yes and no, I guess this connects with Diataxis, we can either have simple instructions or all options listed in one place.

Simple instruction means tutorial. Tutorials are hard to write but crucial for project adoption. Almost every project should have tutorial answering "how do I started" for user and "how do I start" for developers.

All options in Diataxis nomenclature is reference.

What we need are tutorials, so learning oriented content that can be practically used and teach some basics. Other content belongs to next steps of the project since it needs more people understanding projects and we will not win those without tutorials.

> I'm not sure which option is more advanced use case though, there may be a lot of users that don't have TPM2.0 and would prefer to use emulated TPM for testing.

Emulated TPM maybe cool for new tutorial or how to. It depends on complexity. If it is simple step by step to follow to setup emulator and Fobnail for it, then tutorial, if it needs a lot of decisions which will be well described to make this decision wisely it is how to. To sum up I think you maybe writing how-tos.

https://diataxis.fr/tutorials-how-to/#understanding-the-distinction

> GH issues maybe? We would have to prepare a process and a template, probably with some specific logs and maybe certificate chains.

This is good idea, can you create issue for that so we don't forget that planning future work in this project?

> https://github.com/fobnail/fobnail/pull/61 should make it fail with no EK root is passed. Documentation was updated accordingly. Now CI fails as expected

:heart_eyes: I don't have time today to tested it, but I will get back to it as soon as possible.

Reply:
> I have various TPMs in various devices, what would be easiest way to provide feedback?

https://github.com/fobnail/docs/issues/50